File: server/src/collectors/alienvaultOtxJp.js
/**
 * AlienVault OTX (LevelBlue) — threat-intel pulses targeting Japan.
 *
 * Auth: OTX_API_KEY (free, https://otx.alienvault.com/api). The DirectConnect
 * search endpoint accepts free-text queries; we run "japan" and surface the
 * 50 most recent pulses with their indicator counts and adversary tags.
 */

import { createThreatIntelCollector } from '../utils/threatIntelCollectorFactory.js';
import { TOKYO } from './_satelliteSeeds.js';

const URL = 'https://otx.alienvault.com/api/v1/search/pulses?q=japan&limit=50&sort=-modified';
const TIMEOUT_MS = 15000;

export default createThreatIntelCollector({
  sourceId: 'otx_search_pulses',
  description: 'AlienVault OTX — recent pulses matching "japan"',
  envKey: 'OTX_API_KEY',
  envHint: 'Set OTX_API_KEY (free at https://otx.alienvault.com/api)',
  run: async (key) => {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), TIMEOUT_MS);
    const res = await fetch(URL, {
      signal: ctrl.signal,
      headers: { 'X-OTX-API-KEY': key, accept: 'application/json' },
    });
    clearTimeout(t);
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const json = await res.json();
    const pulses = Array.isArray(json?.results) ? json.results : [];
    const features = pulses.map((p, i) => ({
      type: 'Feature',
      geometry: { type: 'Point', coordinates: TOKYO },
      properties: {
        idx: i,
        pulse_id: p.id,
        title: p.name,
        author: p?.author?.username,
        modified: p.modified,
        created: p.created,
        tlp: p.tlp,
        adversary: p.adversary,
        industries: p.industries,
        targeted_countries: p.targeted_countries,
        indicator_count: p.indicator_count,
        tags: p.tags,
        url: p.id ? `https://otx.alienvault.com/pulse/${p.id}` : null,
        source: 'otx_search_pulses',
      },
    }));
    return { features, extraMeta: { query: 'japan' } };
  },
});

File: native/collectors/cyber/sources/alienvault_otx_jp.c
/* collectors/cyber/sources/alienvault_otx_jp.c
 * Port of server/src/collectors/alienvaultOtxJp.js (createThreatIntelCollector).
 * env OTX_API_KEY → OTX pulse search q=japan, 50 recent, TOKYO points. */
#include "../../../source.h"
#include "../../../lib/threatintel.h"
#include "../../../lib/feedlib.h"
#include <stdio.h>
#include <stdlib.h>

#define TOKYO_LON 139.6917
#define TOKYO_LAT 35.6895
#define OTX_URL "https://otx.alienvault.com/api/v1/search/pulses?q=japan&limit=50&sort=-modified"

static void put(cJSON *p, const char *k, cJSON *r, const char *ik) {
  cJSON *v = cJSON_GetObjectItem(r, ik);
  cJSON_AddItemToObject(p, k, v ? cJSON_Duplicate(v, 1) : cJSON_CreateNull());
}

static cJSON *run_fetch(const char *key, const source_ctx *ctx, void *ud) {
  (void)ud;
  char keyh[256];
  snprintf(keyh, sizeof keyh, "X-OTX-API-KEY: %s", key);
  const char *hdrs[] = { keyh, "accept: application/json", NULL };
  cJSON *json = feed_get_json_h(ctx->http, OTX_URL, hdrs, 15000);
  if (!json) return NULL;

  cJSON *results = cJSON_GetObjectItem(json, "results");
  cJSON *features = cJSON_CreateArray();
  int i = 0;
  if (cJSON_IsArray(results)) {
    cJSON *p;
    cJSON_ArrayForEach(p, results) {
      cJSON *f = cJSON_CreateObject();
      cJSON_AddStringToObject(f, "type", "Feature");
      cJSON *g = cJSON_CreateObject();
      cJSON_AddStringToObject(g, "type", "Point");
      cJSON *co = cJSON_CreateArray();
      cJSON_AddItemToArray(co, cJSON_CreateNumber(TOKYO_LON));
      cJSON_AddItemToArray(co, cJSON_CreateNumber(TOKYO_LAT));
      cJSON_AddItemToObject(g, "coordinates", co);
      cJSON_AddItemToObject(f, "geometry", g);

      cJSON *pr = cJSON_CreateObject();        /* EXACT JS key order */
      cJSON_AddNumberToObject(pr, "idx", i);
      put(pr, "pulse_id", p, "id");
      put(pr, "title", p, "name");
      cJSON *auth = cJSON_GetObjectItem(p, "author");
      cJSON *au = auth ? cJSON_GetObjectItem(auth, "username") : NULL;
      cJSON_AddItemToObject(pr, "author",
        au ? cJSON_Duplicate(au, 1) : cJSON_CreateNull());
      put(pr, "modified", p, "modified");
      put(pr, "created", p, "created");
      put(pr, "tlp", p, "tlp");
      put(pr, "adversary", p, "adversary");
      put(pr, "industries", p, "industries");
      put(pr, "targeted_countries", p, "targeted_countries");
      put(pr, "indicator_count", p, "indicator_count");
      put(pr, "tags", p, "tags");
      cJSON *idv = cJSON_GetObjectItem(p, "id");
      if (idv && cJSON_IsString(idv) && idv->valuestring[0]) {
        char url[256];
        snprintf(url, sizeof url, "https://otx.alienvault.com/pulse/%s",
                 idv->valuestring);
        cJSON_AddStringToObject(pr, "url", url);
      } else {
        cJSON_AddItemToObject(pr, "url", cJSON_CreateNull());
      }
      cJSON_AddStringToObject(pr, "source", "otx_search_pulses");
      cJSON_AddItemToObject(f, "properties", pr);
      cJSON_AddItemToArray(features, f);
      i++;
    }
  }
  cJSON_Delete(json);
  return features;
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  int n = threatintel_collect(ctx, sink, "OTX_API_KEY", NULL, run_fetch, NULL);
  return n >= 0 ? 0 : -1;
}

static const source_def alienvault_otx_jp_def = {
  .id = "alienvault-otx-jp", .collector = "cyber",
  .name = "AlienVault OTX (JP-targeted)", .name_ja = "AlienVault OTX (日本標的)",
   .update_interval_sec = 7200, .run = run };
REGISTER_SOURCE(alienvault_otx_jp_def)
